### ISSUE TITLE
Need process for dealing with SDK->API breaking changes

### ISSUE THREAD

**Issue by jeffomatic** [chain/chain#313]
We want to commit changes to the monorepo such that all affected packages can be updated simultaneously. Typically, this means updates to the API are accompanied by corresponding changes in the SDK. However, with commits such as these, it may be impossible to maintain backward compatibility in both of the following situations:

- pre-update SDK -> post-update cored
- post-update SDK -> pre-update cored

This is best explained by example: #306 lays the groundwork for an explicit "retire" action, which is anticipated by various experimental features (entry-based tx structures, etc.). It adds an explicit "retire" action to the API and updates the SDKs accordingly. The change to the API is backward compatible with respect to older versions of the SDK. The change to the SDK is not, however, backward compatible with respect to older versions of the API; it _requires_ the corresponding update to the API.

Unfortunately, the monorepo doesn't live in a vacuum: we have a versioning policy (nascent, but a policy nonetheless) stating that version `x.y.foo` of cored should be compatible with version `x.y.bar` of the SDK. If we adhere to this policy, #306 would motivate an uptick of the minor version. But minor version updates are a grey area: on more than one occasion, we have attached extra political and marketing baggage to the prospect of a minor version update, with the implication that minor version updates cannot be performed casually.

Given the above, minor engineering improvements that happen to break backward compatibility between packages (in some direction, if not bi-directionally) now produce a conundrum: new releases are blocked until we deem the `main` branch worthy of a new minor version. If _build_ version updates are always made on recent commits on the `main` branch, that means we are also blocked on releasing build version updates (i.e., those that have no backward-compatibility issues whatsoever), simply because they arrived on `main` after a breaking change.

Some initial ideas that could break the logjam here:

- Don't commit changes to different packages simultaneously if total backward-compatibility is broken, AND we are not ready for a minor version uptick. This moves away from the monorepo philosophy of updating everything at once, and requires us to have some discipline in remembering to merge deferred updates once a minor version uptick rolls around.
- Relax the version scheme requirement of full-backward compatibility between packages at identical minor versions. For example, version `x.y.foo` of cored must work with any version `x.y.bar` of the SDK, or vice versa, but _not both_. This puts some cognitive load on external developers; we can't reasonably demand that devs remember our version conventions, so we'd have to be super explicit about version requirements in release notes/changelogs. For example, any arbitrary release of the SDK may contain the note "requires version x.y.foo of cored`. There is a burden on external devs to pay very close attention to release notes.
- Use branches for individual package releases, and backport small updates from `main` if it is not feasible to uptick the minor version and the updates occur after a change that breaks backward compatibility for the package in question. This requires some discipline for the branch maintainers to be very aware of the current state of each branch, and to be aware of which features from `main` ought to be included, and which should be deferred.
- Let's just be casual about minor version upticks, and do it whenever we want. This puts the burden on partner leads (or whomever) to communicate this effectively to external devs. There is at least some precedent for external dev teams, in particular of the large enterprise ilk, being nervous (with good reason or not) about minor version instability.

At first glance, I can't say I'm in love with any of the above, so I'm hoping someone with more imagination might have some insight. Thoughts welcome!

**Comment by jeffomatic** [comment]
I had a really great conversation with @devongundry that clarified this issue. The TL;DR is that we should try to lean on the fourth suggestion above. To wit:

> Let's just be casual about minor version upticks, and do it whenever we want.

The caveat about external dev teams being nervous about minor version upticks is, by our current estimate, not a serious concern.
